### PR TITLE
Realize reused Inductor pointwise at computed-read boundary

### LINKED ISSUE #105768: torch.compile uses more memory when using less parameters
### 🐛 Describe the bug

When running a residual block, I noticed that compiled networks that have a 1x1 skip convolution use less memory when applied to all layers rather than just layers where the number of channels change. Note that this is not the case when the activation happens after the convolution. Is this just a quirk of the compile function? I know that pre-activation is not the norm but it seems to have produced better results here: https://towardsdatascience.com/resnet-with-identity-mapping-over-1000-layers-reached-image-classification-bb50a42af03e

Code to reproduce:
```python
import torch.nn.functional as F
import torch.nn as nn
import torch._dynamo
import torch, gc

class ResBlock(nn.Module):
    def __init__(self, in_c: int, nc: int, conv_shortcut: bool = False):
        '''
        in_c: number of input channels
        nc: number of output channels
        '''
        super().__init__()
        self.conv = nn.Conv2d(in_c, nc, 3, padding=1)
        self.skip = nn.Conv2d(in_c, nc, 1) if (in_c != nc or conv_shortcut) else None
    
    def forward(self, x):
        h = self.conv(F.relu(x))
        x = x if self.skip is None else self.skip(x)
        return h + x

device = 'cuda'
for ENABLE_CONV_SHORTCUT in [True, False]:
    print('conv shortcut:', ENABLE_CONV_SHORTCUT)
    with torch.no_grad():
        with torch.amp.autocast('cuda'):
            model = nn.Sequential(
                    ResBlock(256, 256, conv_shortcut=ENABLE_CONV_SHORTCUT),
                    ResBlock(256, 512, conv_shortcut=ENABLE_CONV_SHORTCUT),
                    ResBlock(512, 512, conv_shortcut=ENABLE_CONV_SHORTCUT),
                    ResBlock(512, 1024, conv_shortcut=ENABLE_CONV_SHORTCUT),
                    ResBlock(1024, 1024, conv_shortcut=ENABLE_CONV_SHORTCUT),
                    ResBlock(1024, 2048, conv_shortcut=ENABLE_CONV_SHORTCUT),
                    ResBlock(2048, 2048, conv_shortcut=ENABLE_CONV_SHORTCUT),
            ).to(device)
            model = torch.compile(model)
            print('N params:', sum([p.numel() for p in model.parameters()]))
            x = torch.zeros((8, 256, 64, 64)).to(device)
            for _ in range(2):
                __ = model(x)
                torch.cuda.synchronize()
            print('max mem alloc (GiB):', torch.cuda.max_memory_allocated() / (1024**3))
            print('max mem reser (GiB):', torch.cuda.max_memory_reserved() / (1024**3))

    # clear gpu mem
    del model, x, __
    torch._C._cuda_clearCublasWorkspaces()
    torch._dynamo.reset()
    gc.collect()
    torch.cuda.empty_cache()
    torch.cuda.reset_peak_memory_stats()
    print()
```

Output (OS: Ubuntu 22.04, CPU: i7-12700H, GPU: RTX 3060 Mobile Max-Q):
```
1x1 conv shortcut in all layers: True
N params: 83245568
max mem alloc (GiB): 1.231997013092041
max mem reser (GiB): 1.888671875

1x1 conv shortcut in all layers: False
N params: 77671168
max mem alloc (GiB): 1.3362345695495605
max mem reser (GiB): 2.07421875
```

### Error logs

_No response_

### Minified repro

_No response_

### Versions

PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:20:04) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.19.0-46-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.91
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060 Laptop GPU
Nvidia driver version: 535.54.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          20
On-line CPU(s) list:             0-19
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i7-12700H
CPU family:                      6
Model:                           154
Thread(s) per core:              2
Core(s) per socket:              14
Socket(s):                       1
Stepping:                        3
CPU max MHz:                     4700.0000
CPU min MHz:                     400.0000
BogoMIPS:                        5376.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       544 KiB (14 instances)
L1i cache:                       704 KiB (14 instances)
L2 cache:                        11.5 MiB (8 instances)
L3 cache:                        24 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-19
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.0
[pip3] torch==2.0.1
[pip3] torch-tb-profiler==0.4.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.8.0               h6a678d5_0  
[conda] libmagma                  2.7.1                hc72dce7_3    conda-forge
[conda] libmagma_sparse           2.7.1                hc72dce7_4    conda-forge
[conda] magma                     2.7.1                h2c23e93_0  
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.6           py310h1128e8f_1  
[conda] mkl_random                1.2.2           py310h1128e8f_1  
[conda] numpy                     1.25.0          py310h5f9d8c6_0  
[conda] numpy-base                1.25.0          py310hb5e798b_0  
[conda] pytorch                   2.0.1           py3.10_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h778d358_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-tb-profiler         0.4.1                    pypi_0    pypi
[conda] torchaudio                2.0.2               py310_cu117    pytorch
[conda] torchtriton               2.0.0                     py310    pytorch
[conda] torchvision               0.15.2              py310_cu117    pytorch
[conda] triton                    2.0.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186534

Inductor materializes a reused pointwise expression when its read count is
greater than `realize_reads_threshold`. In #105768, a channel-changing residual
add reads two convolution outputs and two biases, exactly the default threshold
of four, and feeds both the next normalization and the identity skip. Leaving
that add inline keeps both convolution outputs live into the next block and
raises peak CUDA memory for the lower-parameter model.

The previous version multiplied read count by user count, but that over-
materializes cheap input-only fanout such as `y = x + 1` feeding five reductions.
The revised fix handles only graph-reused, non-CPU pointwise expressions exactly
at the threshold, and only when they read at least two distinct registered
computed buffers. Input-only and single-computed-input fanout remain inline;
CPU and above-threshold behavior are unchanged.

Fixes #105768
Generated by my agent

Benchmark Results:
On an NVIDIA B200 using the original seven-block `conv_shortcut=False` repro, a
direct A/B against the legacy implementation after two warmups and over 10 timed
iterations showed:
- Legacy: median 12.477 ms, peak allocated 1.383140 GiB, peak reserved 1.507812 GiB.
- Fix: median 12.761 ms, peak allocated 1.258140 GiB, peak reserved 1.439453 GiB.

This trades about 2.3% runtime for about 9.0% lower peak allocation and 4.5%
lower peak reservation in the memory-OOM repro.

Test Plan:
- `python test/inductor/test_utils.py`
- `python test/inductor/test_cpu_repro.py -k test_cpu_realization_thresholds`
- `lintrunner torch/_inductor/ir.py test/inductor/test_utils.py`
- `lintrunner -a`
- Original CUDA memory repro and legacy-vs-fixed A/B on NVIDIA B200
- CUDA GraphLowering checks for the residual, five-consumer `x + 1`, and
  four-input broadcast cases

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the CPU Inductor CI failures came from applying the new fanout read heuristic to CPU pointwise reuse, which changed expected CPU fusion/codegen structure. I updated the PR to preserve the old CPU read-threshold behavior after the existing CPU heavy-op fast path, while keeping fanout accounting for the CUDA/non-CPU path that fixes #105768.

**Review by yushangdi — CHANGES_REQUESTED** [review]
The fix does work for the issue, but I'm not sure if it's safe to do this overall.

Consider the following example, inlining y is better than materializing it, but under the change in this PR, y will be materialized. 

def workload_new(x):
    y = x + 1
    return (
        y.sum(0),
        y.sum(1),
        y.amax(0),
        y.amax(1),
        (y * y).sum(0),
    )

  Approximate global-memory traffic:

  - Old: each consumer recomputes x + 1 → about 5N reads.
  - New: 1N read of x + 1N write of y + 4–5N reads of y → about 6–7N transfers.

**Comment by jansel** [comment]
My agent says I replaced the fanout multiplier with a narrow liveness rule: only a reused non-CPU pointwise expression exactly at the read threshold and reading at least two registered computed buffers is materialized. Your five-consumer `y = x + 1` example and the related four-input broadcast case now stay inline, while the residual repro still materializes its two convolution outputs; focused tests and B200 lowering checks cover these cases.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
